File: Sources/WordProcessor/Services/SmartQuotes.swift
enum SmartQuotes {
    static func smarten(_ text: String, contextBefore: Character? = nil) -> String {
        guard !text.isEmpty else { return text }

        var result = ""
        result.reserveCapacity(text.count)

        let characters = Array(text)
        for index in characters.indices {
            let character = characters[index]
            let previousCharacter = result.last ?? contextBefore

            if isDoubleQuote(character) {
                let quote = shouldOpenDoubleQuote(
                    in: characters,
                    at: index,
                    previousCharacter: previousCharacter
                ) ? "\u{201c}" : "\u{201d}"
                result.append(quote)
            } else if isSingleQuote(character) {
                let quote = shouldOpenSingleQuote(
                    in: characters,
                    at: index,
                    previousCharacter: previousCharacter
                ) ? "\u{2018}" : "\u{2019}"
                result.append(quote)
            } else {
                result.append(character)
            }
        }

        return result
    }

    private static func shouldOpenDoubleQuote(
        in characters: [Character],
        at index: Int,
        previousCharacter: Character?
    ) -> Bool {
        guard let nextCharacter = nextCharacter(in: characters, at: index) else {
            return isOpeningQuoteContext(previousCharacter)
        }
        guard !nextCharacter.isWhitespace else { return false }

        return isOpeningQuoteContext(previousCharacter)
    }

    private static func shouldOpenSingleQuote(
        in characters: [Character],
        at index: Int,
        previousCharacter: Character?
    ) -> Bool {
        if previousCharacter?.isLetter == true || previousCharacter?.isNumber == true { return false }
        guard let nextCharacter = nextCharacter(in: characters, at: index) else {
            return isOpeningQuoteContext(previousCharacter)
        }
        guard !nextCharacter.isWhitespace else { return false }

        if nextCharacter.isNumber { return false }
        if startsWithApostropheElision(characters.suffix(from: index + 1)) { return false }

        return isOpeningQuoteContext(previousCharacter)
    }

    private static func nextCharacter(in characters: [Character], at index: Int) -> Character? {
        let nextIndex = index + 1
        guard nextIndex < characters.count else { return nil }
        return characters[nextIndex]
    }

    private static func isDoubleQuote(_ character: Character) -> Bool {
        character == "\"" || character == "\u{201c}" || character == "\u{201d}"
    }

    private static func isSingleQuote(_ character: Character) -> Bool {
        character == "'" || character == "\u{2018}" || character == "\u{2019}"
    }

    private static func isOpeningQuoteContext(_ character: Character?) -> Bool {
        guard let character else { return true }
        if character.isWhitespace { return true }
        if "([{<-\u{2013}\u{2014}\u{201c}\u{2018}".contains(character) { return true }
        return false
    }

    private static func startsWithApostropheElision(_ suffix: ArraySlice<Character>) -> Bool {
        let text = String(suffix).lowercased()
        if text.range(of: #"^[a-z]'"#, options: .regularExpression) != nil {
            return true
        }

        return [
            "tis",
            "twas",
            "twere",
            "cause",
            "cuz",
            "em",
            "til",
            "bout",
            "round",
        ].contains { text.hasPrefix($0) }
    }
}

File: Editor/src/smartQuotes.ts
import { Editor, Extension } from '@tiptap/core';
import { Plugin, PluginKey } from '@tiptap/pm/state';
import { isAlphaNumericCharacter, isWhitespaceCharacter, lastCharacter } from './utils';

const smartQuotesPluginKey = new PluginKey('smartQuotes');
const SMART_QUOTES_TRANSACTION_META = 'smartQuotesNormalized';

let smartQuotesNormalizationFrame: number | null = null;

function isOpeningQuoteContext(character: string): boolean {
  return !character || isWhitespaceCharacter(character) || /[\([{<\u2013\u2014-]/.test(character) || character === '\u201C' || character === '\u2018';
}

function startsWithApostropheElision(text: string): boolean {
  const lower = text.toLowerCase();
  if (/^[a-z]'/.test(lower)) return true;
  return [
    'tis',
    'twas',
    'twere',
    'cause',
    'cuz',
    'em',
    'til',
    'bout',
    'round',
  ].some((prefix) => lower.startsWith(prefix));
}

function shouldOpenDoubleQuote(text: string, index: number, previousCharacter: string): boolean {
  const nextCharacter = text[index + 1] || '';
  if (!nextCharacter || isWhitespaceCharacter(nextCharacter)) return false;
  return isOpeningQuoteContext(previousCharacter);
}

function shouldOpenSingleQuote(text: string, index: number, previousCharacter: string): boolean {
  const nextCharacter = text[index + 1] || '';
  if (!nextCharacter || isWhitespaceCharacter(nextCharacter)) return false;
  if (isAlphaNumericCharacter(previousCharacter)) return false;
  if (/[0-9]/.test(nextCharacter)) return false;
  if (startsWithApostropheElision(text.slice(index + 1))) return false;
  return isOpeningQuoteContext(previousCharacter);
}

export function smartifyQuotes(text: string): string {
  let result = '';

  for (let index = 0; index < text.length; index += 1) {
    const character = text[index];
    const previousCharacter = lastCharacter(result);

    if (character === '"') {
      result += shouldOpenDoubleQuote(text, index, previousCharacter) ? '\u201C' : '\u201D';
    } else if (character === "'") {
      result += shouldOpenSingleQuote(text, index, previousCharacter) ? '\u2018' : '\u2019';
    } else {
      result += character;
    }
  }

  return result;
}

export function smartifyQuotesWithContext(text: string, contextBefore = ''): string {
  if (!text) return text;
  const syntheticPrefix = contextBefore || ' ';
  return smartifyQuotes(`${syntheticPrefix}${text}`).slice(syntheticPrefix.length);
}

export function contextCharacterBefore(doc: any, pos: number): string {
  if (pos <= 0) return '';
  return doc.textBetween(Math.max(pos - 1, 0), pos, '', '');
}

function buildSmartQuotesNormalizationTransaction(state: any) {
  let transaction = state.tr;
  let changed = false;

  state.doc.descendants((node: any, pos: number) => {
    if (!node.isText || typeof node.text !== 'string') return;
    if (!node.text.includes('"') && !node.text.includes("'")) return;

    const converted = smartifyQuotesWithContext(
      node.text,
      contextCharacterBefore(state.doc, pos)
    );

    if (converted === node.text) return;

    transaction = transaction.replaceWith(
      pos,
      pos + node.nodeSize,
      state.schema.text(converted, node.marks)
    );
    changed = true;
  });

  if (!changed) {
    return null;
  }

  transaction.setMeta(SMART_QUOTES_TRANSACTION_META, true);
  return transaction;
}

export function normalizeDocumentSmartQuotes(editor: Editor): boolean {
  const transaction = buildSmartQuotesNormalizationTransaction(editor.state);
  if (!transaction) {
    return false;
  }

  editor.view.dispatch(transaction);
  return true;
}

export function scheduleSmartQuotesNormalization(editor: Editor) {
  if (smartQuotesNormalizationFrame !== null) {
    window.cancelAnimationFrame(smartQuotesNormalizationFrame);
  }

  smartQuotesNormalizationFrame = window.requestAnimationFrame(() => {
    smartQuotesNormalizationFrame = null;
    normalizeDocumentSmartQuotes(editor);
  });
}

/**
 * Apply smart quotes to all text nodes in a DOM tree (preserving HTML structure).
 */
export function smartifyDOMTextNodes(root: Node, contextBefore = ''): void {
  const ownerDocument = root.ownerDocument ?? document;
  const walker = ownerDocument.createTreeWalker(root, NodeFilter.SHOW_TEXT);
  const textNodes: Text[] = [];
  let previousText = contextBefore;
  while (walker.nextNode()) {
    textNodes.push(walker.currentNode as Text);
  }
  for (const node of textNodes) {
    const converted = smartifyQuotesWithContext(node.data, lastCharacter(previousText));
    if (converted !== node.data) {
      node.data = converted;
    }
    previousText = converted;
  }
}

export function smartifyHTMLFragment(html: string, contextBefore = ''): string {
  const parsed = new DOMParser().parseFromString(html, 'text/html');
  smartifyDOMTextNodes(parsed.body, contextBefore);
  return parsed.body.innerHTML;
}

function singleElementChildIgnoringWhitespace(root: HTMLElement): HTMLElement | null {
  let element: HTMLElement | null = null;

  for (const child of Array.from(root.childNodes)) {
    if (child.nodeType === Node.TEXT_NODE) {
      if (child.textContent?.trim()) return null;
      continue;
    }

    if (child instanceof HTMLElement) {
      if (element) return null;
      element = child;
      continue;
    }

    return null;
  }

  return element;
}

function unwrapSingleParagraphHTML(html: string): string {
  const parsed = new DOMParser().parseFromString(html, 'text/html');
  const onlyChild = singleElementChildIgnoringWhitespace(parsed.body);
  if (!onlyChild) return html;

  const tagName = onlyChild.tagName.toLowerCase();
  if (tagName !== 'p' && tagName !== 'div') return html;

  return onlyChild.innerHTML;
}

function isTextblockRange(ed: Editor, from: number, to: number): boolean {
  try {
    const resolvedFrom = ed.state.doc.resolve(from);
    const resolvedTo = ed.state.doc.resolve(Math.max(from, to));
    return resolvedFrom.sameParent(resolvedTo) && resolvedFrom.parent.isTextblock;
  } catch (_) {
    return false;
  }
}

export function prepareReplacementHTMLForRange(
  ed: Editor,
  from: number,
  to: number,
  html: string
): string {
  const smartified = smartifyHTMLFragment(html, contextCharacterBefore(ed.state.doc, from));
  return isTextblockRange(ed, from, to)
    ? unwrapSingleParagraphHTML(smartified)
    : smartified;
}

export const SmartQuotes = Extension.create({
  name: 'smartQuotes',

  addProseMirrorPlugins() {
    return [
      new Plugin({
        key: smartQuotesPluginKey,
        props: {
          handleTextInput(view, from, to, text) {
            const converted = smartifyQuotesWithContext(
              text,
              contextCharacterBefore(view.state.doc, from)
            );

            if (converted === text) {
              return false;
            }

            view.dispatch(view.state.tr.insertText(converted, from, to));
            return true;
          },
        },
        appendTransaction(transactions, _oldState, newState) {
          if (!transactions.some((transaction) => transaction.docChanged)) {
            return null;
          }

          if (transactions.some((transaction) => transaction.getMeta(SMART_QUOTES_TRANSACTION_META))) {
            return null;
          }
          return buildSmartQuotesNormalizationTransaction(newState);
        },
      }),
    ];
  },
});
